src/impl_owned_array.rs

```diff
@@ -657,6 +687,43 @@ impl<A, D> Array<A, D>
 
         Ok(())
     }
+
+    /// Reserve capacity to grow array along `axis` by at least `additional` elements.
+    ///
+    /// Existing elements of `array` are untouched and the backing storage is grown by
+    /// calling the underlying `reserve` method of the `OwnedRepr`.
+    ///
+    /// This is useful when pushing or appending repeatedly to an array to avoid multiple
+    /// allocations.
+    ///
+    /// ```rust
+    /// use ndarray::{Array3, Axis};
+    /// let mut a = Array3::<i32>::zeros((0,2,4));
+    /// a.reserve(Axis(0), 1000);
+    /// assert!(a.into_raw_vec().capacity() >= 2*4*1000);
+    /// ```
+    pub fn reserve(&mut self, axis: Axis, additional: usize)
+    where
+        D: RemoveAxis,
+    {
+        let self_dim = self.raw_dim();
```

Comment:
I think this method should have an assert that axis is in bounds, that len_to_append didn't overflow (and if data.reserve doesn't check it, also if the calculation in there doesn't overflow)

Then document the panic on invalid argument as is usual (see other methods that can panic)

Reply:
Good point - I think the latest commits should address this. `data.reserve` does check whether the new capacity would be larger than `isize::MAX`, but I've added some bounds checking anyway that returns a ShapeError in that case in keeping with other functions (in addition to the panic if `len_to_append` would overflow `usize::MAX` or if adding it to the current len causes it to wrap).